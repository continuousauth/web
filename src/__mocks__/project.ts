import { FullProject } from '../common/types';

export const mockProject = (): FullProject => ({
  secret: 'my_secret',
  enabled: true,
  requester_circleCI: null,
  requester_gitHub: null,
  responder_slack: null,
  id: '123',
  repoName: 'my-repo',
  repoOwner: 'my-owner',
  defaultBranch: 'main',
});
