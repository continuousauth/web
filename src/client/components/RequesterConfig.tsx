import * as React from 'react';
import { Alert, Heading, Pane, Paragraph, Tab, Tablist } from 'evergreen-ui';

import { FullProject } from '../../common/types';

import styles from './ReqResConfig.scss';
import { CircleCILogo } from './icons/CircleCI';
import { CircleCIRequesterConfig } from './configurators/CircleCIRequesterConfig';
import { TravisCILogo } from './icons/TravisCI';
import { TravisCIRequesterConfig } from './configurators/TravisCIRequesterConfig';

export interface Props {
  project: FullProject;
  setProject: (newProject: FullProject) => void;
}

enum RequesterTab {
  NOTHING_YET,
  CIRCLE_CI,
  TRAVIS_CI,
}

const defaultTabForProject = (project: FullProject) => {
  if (project.requester_circleCI) return RequesterTab.CIRCLE_CI;
  if (project.requester_travisCI) return RequesterTab.TRAVIS_CI;
  return RequesterTab.NOTHING_YET;
};

export function RequesterConfig({ project, setProject }: Props) {
  const [activeTab, setActiveTab] = React.useState(RequesterTab.NOTHING_YET);
  React.useEffect(() => {
    setActiveTab(defaultTabForProject(project));
  }, [defaultTabForProject(project)]);

  const [showRequesterHelp, setShowRequesterHelp] = React.useState(
    defaultTabForProject(project) === RequesterTab.NOTHING_YET,
  );
  React.useEffect(() => {
    if (defaultTabForProject(project) === RequesterTab.NOTHING_YET && !showRequesterHelp) {
      setShowRequesterHelp(true);
    }
  }, [project]);

  return (
    <Pane>
      <Heading marginBottom={8}>Requester</Heading>
      <Pane>
        {showRequesterHelp ? (
          <Alert
            intent="none"
            title="What is a Requester?"
            isRemoveable
            onRemove={() => setShowRequesterHelp(false)}
          >
            A Requester is how your automation asks CFA for a 2FA token. This is normally your CI
            provider, E.g. Circle CI.
          </Alert>
        ) : null}
      </Pane>
      <Pane marginY={8}>
        <Tablist>
          <Tab
            onSelect={() => setActiveTab(RequesterTab.CIRCLE_CI)}
            isSelected={activeTab === RequesterTab.CIRCLE_CI}
            style={{ paddingLeft: 28, position: 'relative' }}
          >
            <CircleCILogo className={styles.tabIcon} /> Circle CI
          </Tab>
          <Tab
            onSelect={() => setActiveTab(RequesterTab.TRAVIS_CI)}
            isSelected={activeTab === RequesterTab.TRAVIS_CI}
            style={{ paddingLeft: 28, position: 'relative' }}
          >
            <TravisCILogo className={styles.tabIcon} /> Travis CI
          </Tab>
          <Tab disabled>More Coming Soon...</Tab>
        </Tablist>
      </Pane>
      <Pane marginY={8} className={styles.configBox}>
        {activeTab === RequesterTab.NOTHING_YET ? (
          <Paragraph>No Requester has been configured, choose one to get started!</Paragraph>
        ) : activeTab === RequesterTab.CIRCLE_CI ? (
          <CircleCIRequesterConfig project={project} setProject={setProject} />
        ) : activeTab === RequesterTab.TRAVIS_CI ? (
          <TravisCIRequesterConfig project={project} setProject={setProject} />
        ) : null}
      </Pane>
    </Pane>
  );
}
