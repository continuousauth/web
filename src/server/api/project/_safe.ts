import * as crypto from 'crypto';
import * as express from 'express';

import { Project } from '../../db/models';
import { ExpressRequest, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';

export const generateNewSecret = (size: number) => {
  if (size > 256) throw new Error('size must be <= 256');
  return crypto.randomBytes(256).toString('hex').slice(0, size);
};

export const sanitizeProject = (project: Project) => project;

export const getProjectFromIdAndCheckPermissions = async (
  id: string,
  req: ExpressRequest,
  res: express.Response,
) => {
  const project = await Project.findOne({
    where: {
      id: id,
      enabled: true,
    },
    include: Project.allIncludes,
  });
  // Project not existing and user not having permission should look identical
  // so that folks can't sniff which projects are active on CFA
  if (!project) {
    res.status(401).json({
      error: 'Current user is not an admin of the target repository',
    });
    return null;
  }

  if (!(await hasAdminAccessToTargetRepo(req, project.id))) {
    res.status(401).json({
      error: 'Current user is not an admin of the target repository',
    });
    return null;
  }

  return project;
};
