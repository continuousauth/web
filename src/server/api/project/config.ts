import axios from 'axios';
import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';

import { validate, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';
import { createA } from '../../helpers/a';
import {
  Project,
  SlackResponderLinker,
  withTransaction,
  CircleCIRequesterConfig,
  TravisCIRequesterConfig,
} from '../../db/models';
import { getProjectFromIdAndCheckPermissions } from './_safe';

const d = debug('cfa:server:api:project:config');
const a = createA(d);

export async function updateCircleEnvVars(project: Project, accessToken: string) {
  const client = axios.create({
    baseURL: 'https://circleci.com/api/v1.1',
    auth: {
      username: accessToken,
      password: '',
    },
    validateStatus: () => true,
  });

  const existing = await client.get(
    `/project/github/${project.repoOwner}/${project.repoName}/envvar`,
  );
  if (existing.status !== 200) return;

  if (existing.data.find(item => item.name === 'CFA_SECRET')) {
    await client.delete(
      `/project/github/${project.repoOwner}/${project.repoName}/envvar/CFA_SECRET`,
    );
  }
  await client.post(`/project/github/${project.repoOwner}/${project.repoName}/envvar`, {
    name: 'CFA_SECRET',
    value: project.secret,
  });

  if (existing.data.find(item => item.name === 'CFA_PROJECT_ID')) {
    await client.delete(
      `/project/github/${project.repoOwner}/${project.repoName}/envvar/CFA_PROJECT_ID`,
    );
  }
  await client.post(`/project/github/${project.repoOwner}/${project.repoName}/envvar`, {
    name: 'CFA_PROJECT_ID',
    value: project.id,
  });
}

export function configRoutes() {
  const router = express();

  router.post(
    '/:id/config/requesters/circleci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://circleci.com/api/v1.1/project/gh/${project.repoOwner}/${project.repoName}/checkout-key`,
          {
            auth: {
              username: req.body.accessToken,
              password: '',
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on CircleCI',
          });
        }

        await updateCircleEnvVars(project, req.body.accessToken);

        const newProject = await withTransaction(async t => {
          const config = await CircleCIRequesterConfig.create(
            {
              accessToken: req.body.accessToken,
            },
            {
              returning: true,
            },
          );
          await project.resetAllRequesters(t);
          project.requester_circleCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/requesters/travisci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://api.travis-ci.org/repo/${project.repoOwner}%2F${project.repoName}`,
          {
            headers: {
              'Travis-API-Version': '3',
              Authorization: `token ${req.body.accessToken}`,
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on Travis CI',
          });
        }

        const newProject = await withTransaction(async t => {
          const config = await TravisCIRequesterConfig.create(
            {
              accessToken: req.body.accessToken,
            },
            {
              returning: true,
            },
          );
          await project.resetAllRequesters(t);
          project.requester_travisCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const linker = await withTransaction(async t => {
          await SlackResponderLinker.destroy({
            where: {
              projectId: project.id,
            },
            transaction: t,
          });
          return await SlackResponderLinker.create(
            {
              projectId: project.id,
            },
            {
              transaction: t,
              returning: true,
            },
          );
        });

        res.json({
          linker,
          slackClientId: process.env.SLACK_CLIENT_ID,
        });
      },
    ),
  );

  router.patch(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          usernameToMention: Joi.string()
            .min(1)
            .max(50)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        if (!project.responder_slack) {
          return res.status(400).json({
            error: 'Project is not configured to use Slack as a responder',
          });
        }

        project.responder_slack.usernameToMention = req.body.usernameToMention;
        await project.responder_slack.save();

        res.json(project);
      },
    ),
  );

  return router;
}
