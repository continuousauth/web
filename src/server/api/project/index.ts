import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';
import * as Octokit from '@octokit/rest';

import { createA } from '../../helpers/a';
import { validate, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';
import { Project, withTransaction, OTPRequest } from '../../db/models';
import { configRoutes } from './config';
import { sanitizeProject, generateNewSecret, getProjectFromIdAndCheckPermissions } from './_safe';

const d = debug('cfa:server:api:auth');
const a = createA(d);

export function projectRoutes() {
  const router = express();

  router.post(
    '/',
    validate(
      {
        a,
        body: {
          repoId: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        if (!(await hasAdminAccessToTargetRepo(req, `${req.body.repoId}`))) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        const github = new Octokit({
          auth: req.user.accessToken,
        });

        const repoResponse = await github.request('GET /repositories/:id', {
          id: req.body.repoId,
        });
        const repo: Octokit.ReposGetResponse = repoResponse.data;

        const existingProject = await Project.findByPk(req.body.repoId);
        if (existingProject) {
          existingProject.repoName = repo.name;
          existingProject.repoOwner = repo.owner.login;
          // If the project existed but was disabled let's enable it and
          // generate a new secret so that the old info is invalid.
          if (!existingProject.enabled) {
            existingProject.enabled = true;
            existingProject.secret = generateNewSecret(256);
          }
          await existingProject.save();
          return res.status(200).json(existingProject);
        }

        const project = await Project.create({
          id: req.body.repoId,
          repoName: repo.name,
          repoOwner: repo.owner.login,
          secret: generateNewSecret(256),
          defaultBranch: repo.default_branch,
        });
        res.status(201).json(project);
      },
    ),
  );

  router.get(
    '/:id',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await Project.findOne({
          where: {
            id: req.params.id,
            enabled: true,
          },
          include: Project.allIncludes,
        });
        // Project not existing and user not having permission should look identical
        // so that folks can't sniff which projects are active on CFA
        if (!project) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        if (!(await hasAdminAccessToTargetRepo(req, project.id))) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.get(
    '/:id/log',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        res.json(
          (
            await OTPRequest.findAll({
              where: {
                projectId: project.id,
              },
            })
          )
            .map(req => {
              const simpleReq = req.get() as OTPRequest;
              delete simpleReq.proof;
              delete simpleReq.requestMetadata;
              delete simpleReq.responseMetadata;
              return simpleReq;
            })
            .sort((a, b) => b.requested.getTime() - a.requested.getTime()),
        );
      },
    ),
  );

  router.patch(
    '/:id/secret',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        project.secret = generateNewSecret(256);
        await project.save();

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.delete(
    '/:id',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        await withTransaction(async t => {
          project.enabled = false;
          await project.resetAllResponders(t);
          await project.resetAllRequesters(t);
          await project.save({ transaction: t });
        });

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.use(configRoutes());

  return router;
}
