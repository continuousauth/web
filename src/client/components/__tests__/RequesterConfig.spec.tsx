/**
 * @jest-environment jsdom
 */

jest.mock('../configurators/CircleCIRequesterConfig', () => ({
  CircleCIRequesterConfig: () => null,
}));

jest.mock('../configurators/TravisCIRequesterConfig', () => ({
  TravisCIRequesterConfig: () => null,
}));

import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { RequesterConfig } from '../RequesterConfig';
import { mockProject } from '../../../__mocks__/project';

describe('<RequesterConfig />', () => {
  it('Should render', () => {
    const setProject = jest.fn();
    const wrapper = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('Should render a help box by default', () => {
    const setProject = jest.fn();
    const mounted = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    expect(mounted.find('Memo(ForwardRef(Alert))')).toHaveLength(1);
  });

  it('Should hide the help box when it is closed', () => {
    const setProject = jest.fn();
    const mounted = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    act(() => (mounted.find('Memo(ForwardRef(Alert))').prop('onRemove') as Function)());
    mounted.setProps({});
    expect(mounted.find('Memo(ForwardRef(Alert))')).toHaveLength(0);
  });

  it.skip('Should reopen the help box for a new project if it is closed', () => {
    const setProject = jest.fn();
    const mounted = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    act(() => (mounted.find('Memo(ForwardRef(Alert))').prop('onRemove') as Function)());
    mounted.setProps({});
    expect(mounted.find('Memo(ForwardRef(Alert))')).toHaveLength(0);
    mounted.setProps({
      project: mockProject(),
    });
    mounted.setProps({});
    expect(mounted.find('Memo(ForwardRef(Alert))')).toHaveLength(1);
  });

  it('should ask us to configure a requester if none is configured', () => {
    const setProject = jest.fn();
    const project = mockProject();
    const mounted = shallow(<RequesterConfig project={project} setProject={setProject} />);
    expect(
      mounted
        .find('Memo(ForwardRef(Paragraph))')
        .at(0)
        .text()
        .includes('choose one'),
    ).toBeTruthy();
  });

  it('Should show the circleci configurator when that tab is selected', () => {
    const setProject = jest.fn();
    const mounted = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    expect(mounted.find('CircleCIRequesterConfig')).toHaveLength(0);
    act(() =>
      (mounted
        .find('Memo(ForwardRef(Tab))')
        .at(0)
        .prop('onSelect') as Function)(),
    );
    mounted.setProps({});
    expect(mounted.find('CircleCIRequesterConfig')).toHaveLength(1);
  });

  it('Should show the travisci configurator when that tab is selected', () => {
    const setProject = jest.fn();
    const mounted = shallow(<RequesterConfig project={mockProject()} setProject={setProject} />);
    expect(mounted.find('TravisCIRequesterConfig')).toHaveLength(0);
    act(() =>
      (mounted
        .find('Memo(ForwardRef(Tab))')
        .at(1)
        .prop('onSelect') as Function)(),
    );
    mounted.setProps({});
    expect(mounted.find('TravisCIRequesterConfig')).toHaveLength(1);
  });

  it('Should show the circleci configurator when circleci has been configured on the provided project', () => {
    const setProject = jest.fn();
    const project = mockProject();
    project.requester_circleCI = {
      accessToken: 'test123',
    };
    const mounted = mount(<RequesterConfig project={project} setProject={setProject} />);
    expect(mounted.find('CircleCIRequesterConfig')).toHaveLength(1);
  });

  it('Should show the travisci configurator when travisci has been configured on the provided project', () => {
    const setProject = jest.fn();
    const project = mockProject();
    project.requester_travisCI = {
      accessToken: 'test123',
    };
    const mounted = mount(<RequesterConfig project={project} setProject={setProject} />);
    expect(mounted.find('TravisCIRequesterConfig')).toHaveLength(1);
  });
});
