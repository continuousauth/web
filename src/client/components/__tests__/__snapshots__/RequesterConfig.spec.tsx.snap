// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequesterConfig /> Should render 1`] = `
<Memo(ForwardRef(Pane))>
  <Memo(ForwardRef(Heading))
    marginBottom={8}
  >
    Requester
  </Memo(ForwardRef(Heading))>
  <Memo(ForwardRef(Pane))>
    <Memo(ForwardRef(Alert))
      intent="none"
      isRemoveable={true}
      onRemove={[Function]}
      title="What is a Requester?"
    >
      A Requester is how your automation asks CFA for a 2FA token. This is normally your CI provider, E.g. Circle CI.
    </Memo(ForwardRef(Alert))>
  </Memo(ForwardRef(Pane))>
  <Memo(ForwardRef(Pane))
    marginY={8}
  >
    <ForwardRef(Tablist)>
      <Memo(ForwardRef(Tab))
        isSelected={false}
        onSelect={[Function]}
        style={
          {
            "paddingLeft": 28,
            "position": "relative",
          }
        }
      >
        <CircleCILogo
          className="tabIcon"
        />
         Circle CI
      </Memo(ForwardRef(Tab))>
      <Memo(ForwardRef(Tab))
        isSelected={false}
        onSelect={[Function]}
        style={
          {
            "paddingLeft": 28,
            "position": "relative",
          }
        }
      >
        <TravisCILogo
          className="tabIcon"
        />
         Travis CI
      </Memo(ForwardRef(Tab))>
      <Memo(ForwardRef(Tab))
        disabled={true}
      >
        More Coming Soon...
      </Memo(ForwardRef(Tab))>
    </ForwardRef(Tablist)>
  </Memo(ForwardRef(Pane))>
  <Memo(ForwardRef(Pane))
    className="configBox"
    marginY={8}
  >
    <Memo(ForwardRef(Paragraph))>
      No Requester has been configured, choose one to get started!
    </Memo(ForwardRef(Paragraph))>
  </Memo(ForwardRef(Pane))>
</Memo(ForwardRef(Pane))>
`;
